alacritty_terminal/src/locale.rs

```diff
@@ -12,22 +12,62 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 #![allow(clippy::let_unit_value)]
-#![cfg(target_os = "macos")]
-use libc::{setlocale, LC_CTYPE};
+
 use std::env;
 use std::ffi::{CStr, CString};
 use std::os::raw::c_char;
-use std::ptr::null;
 use std::slice;
 use std::str;
 
+use libc::{setlocale, LC_ALL, LC_CTYPE};
+use log::debug;
 use objc::runtime::{Class, Object};
 
+const FALLBACK_LOCALE: &str = "UTF-8";
+
 pub fn set_locale_environment() {
-    let locale_id = unsafe {
+    let env_locale_c = CString::new("").unwrap();
+    let env_locale_ptr = unsafe { setlocale(LC_ALL, env_locale_c.as_ptr()) };
+    if !env_locale_ptr.is_null() {
+        let env_locale = unsafe { CStr::from_ptr(env_locale_ptr).to_string_lossy() };
+
+        // Assume `C` locale means unchanged, since it is the default anyways
+        if env_locale != "C" {
+            debug!("Using environment locale: {}", env_locale);
+            return;
+        }
+    }
+
+    let system_locale = system_locale();
+
+    // Set locale to system locale
+    let system_locale_c = CString::new(system_locale.clone()).expect("nul byte in system locale");
+    let lc_all = unsafe { setlocale(LC_ALL, system_locale_c.as_ptr()) };
+
+    // Check if system locale was valid or not
+    if lc_all.is_null() {
+        // Use fallback locale
+        debug!("Using fallback locale: {}", FALLBACK_LOCALE);
+
+        let fallback_locale_c = CString::new(FALLBACK_LOCALE).unwrap();
+        unsafe { setlocale(LC_CTYPE, fallback_locale_c.as_ptr()) };
+
+        env::set_var("LC_CTYPE", FALLBACK_LOCALE);
+    } else {
+        // Use system locale
+        debug!("Using system locale: {}", system_locale);
+
+        env::set_var("LC_ALL", system_locale);
```

Comment:
We should consider replacing the `LC_ALL` instances here with `LANG`, since that's probably more commonly used and checked against.

If we use `LANG`, then `LC_ALL` will be empty. If we use `LC_ALL`, then `LANG` will be empty. So it's probably better to have `LANG` filled. All other values like `LC_CTYPE` are automatically set when either is present.

Reply:
I belive this is the right approach.

Reply:
Let's go for it then.